### ISSUE TITLE
Missing headers in RestTemplate http requests using OpenTelemetry Spring Auto-Configuration

### ISSUE THREAD

**Issue by Xitric** [open-telemetry/opentelemetry-java-instrumentation#2624]
**Describe the bug**
We have observed, that when using OpenTelemetry Spring Auto-Configuration, http requests created using `RestTemplate` do not contain context propagation headers in version 1.0.1-alpha. However, when downgrading to version 0.16.0, the headers are correctly injected.

We have also tried to manually use the Global tracer, acquired with `GlobalOpenTelemetry#getTracer(String)`, to inject headers into outbound requests. This injection process is working as expected. Therefore, we believe there might be an issue in the Auto-Configuration module.

**Steps to reproduce**
We have created a [small repository](https://github.com/em135/OtelSpringBoot), which contains tests showing that headers are correctly injected in version 0.16.0, but missing in version 1.0.1:

**What did you expect to see?**
We expected that http requests sent using RestTemplate from a Spring Boot application would contain context propagation headers.

**What did you see instead?**
In version 1.0.1, the outbound request contains no context propagation headers. In fact, the headers are the same as if not using the Auto-Configuration module at all.

**What version are you using?**
Version 1.0.1-alpha of the Auto-Configuration module, and version 1.0.1 of everything else. Precise dependencies are:

- io.opentelemetry.instrumentation:opentelemetry-spring-boot-autoconfigure:1.0.1-alpha
- io.opentelemetry:opentelemetry-api:1.0.1
- io.opentelemetry:opentelemetry-exporter-logging:1.0.1
- io.opentelemetry:opentelemetry-context:1.0.1

For Spring Boot, we are using these dependencies:

- org.springframework:spring-web:5.3.5
- org.springframework.boot:spring-boot-autoconfigure:2.4.4

These dependencies can also be seen in the repository linked above.

**Environment**
Compiler: Oracle Java SE Development Kit 15.0.2
OS: Windows 10 Home

**Additional context**
We use the following JVM arguments when running the test from Maven:

```
-javaagent:opentelemetry-javaagent-all.jar -Dotel.resource.attributes=service.name=demo -Dotel.traces.exporter=none -Dotel.metrics.exporter=none -Dotel.propagators=b3multi
```

We have disabled exporters and enabled B3 multi header propagation, because we intend to use this agent together with Istio on OpenShift.

**Comment by abhi005** [comment]
hey @Xitric I would like to work on this issue.

**Comment by Xitric** [comment]
Feel free to dive in @abhi005.

I also opened this issue to find out whether this is expected behavior (maybe I had configured something wrong in the new version), or if there does indeed seem to be an issue in the Auto-Configuration module. I would normally have attempted to solve the issue myself, but I am currently preoccupied with other projects.

**Comment by abhi005** [comment]
Cool.
Actually this would be my first time contributing and its a mandatory requirement of my course work as well so it would be helpfull if you can suggest me a starting point. from where to start looking.

**Comment by Xitric** [comment]
Unfortunately, my knowledge of the Auto-Configuration module is limited. However, I would personally look at the [HttpClientTracer](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/HttpClientTracer.java), which seems to be the supertype of the tracer used by the [RestTemplateInterceptor](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-web-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/httpclients/RestTemplateInterceptor.java#L18). It is the tracer, that eventually initiates the injection of headers inside the [startSpan()](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/2a2d4ae2f844b2c5dbb61a5d566a9a9fadeb19aa/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/HttpClientTracer.java#L75) method, but contrary to my expectation, no headers are actually being inserted. I do not know the cause.

**Comment by trask** [comment]
hey @Xitric, we are collaborating with the spring team over at https://github.com/spring-cloud-incubator/spring-cloud-sleuth-otel, I'm curious if that project addresses your needs?

**Comment by Xitric** [comment]
Thanks for the reply @trask! Yes, I believe that project meets our requirements. If the Spring instrumentation in this project is being replaced in favor of the Spring Cloud Sleuth project, feel free to close this issue. Although, for the time being, I would still consider the behavior we experienced to be a bug, unless I am mistaken.

**Comment by trask** [comment]
Hey, I tried out your repro (thanks!).

I noticed the tests are also using the javaagent (in addition to opentelemetry-spring-boot-autoconfigure).

If you are using the javaagent, you shouldn't need opentelemetry-spring-boot-autoconfigure, and trace propagation should be working (your test still fails with the javaagent because you are mocking RestTemplate, while the javaagent instruments the underlying http client used by RestTemplate).

In general, opentelemetry-spring-boot-autoconfigure is designed as an alternative to the javaagent. The configuration that you do via opentelemetry-spring-boot-autoconfigure affects the OpenTelemetry SDK on your classpath, and that configuration will not sync over to the OpenTelemetry SDK that is embedded in the javaagent.
